JsonLedgerApiSample/src/main/java/com/example/store/IntegrationStore.java

```diff
@@ -24,6 +28,170 @@ public class IntegrationStore {
     private String sourceSynchronizerId = null;
     private String lastIngestedRecordTime = null;
     private final String treasuryParty;
+    private final HashMap<String, Balances> userBalances = new HashMap<>();
+
+    public BigDecimal getBalance(InstrumentId instrumentId, String depositId) {
+        return userBalances.getOrDefault(depositId, new Balances()).getBalance(instrumentId);
+    }
+
+    private record TransferInfo(
+            String sender,
+            String depositId) {
+    }
+
+    private class TransactionParser {
+        static final String SPLICE_PREFIX = "splice.lfdecentralizedtrust.org/";
+        static final String REASON_KEY = SPLICE_PREFIX + "reason";
+        static final String SENDER_KEY = SPLICE_PREFIX + "sender";
+        static final String TRANSFER_KIND_KEY = SPLICE_PREFIX + "tx-kind";
+
+        // Generic parsing of metadata is only required if the packages use custom transfer workflows.
+        // It also requires the package to set the metadata in the expected format.
+        // To avoid confusion, we whitelist the package names where generic metadata parsing is enabled.
+        //
+        // For all others we expect to use the token standard choices to parse the transfers.
+        static final Set<String> packagesWithGenericMetadataParsing = Set.of("splice-amulet");
+
+        private final Integer lastDescendantNodeId;
+        private final Iterator<Event> events;
+        private TransferInfo transferInfo;
+
+        /**
+         * Parse the events in the iterator up to and including the event with the given node ID.
+         * <p>
+         * This parser properly follows the tree structure of exercise nodes and their children.
+         * It will always parse at least one event.
+         */
+        TransactionParser(Iterator<Event> events, int lastDescendantNodeId, TransferInfo transferInfo) {
+            assert events.hasNext();
+            this.events = events;
+            this.lastDescendantNodeId = lastDescendantNodeId;
+            this.transferInfo = transferInfo;
+        }
+
+        /**
+         * Parse the transaction and return the node ID of the last event parsed.
+         */
+        int parse() {
+            assert events.hasNext();
+            int nodeId;
+            do {
+                nodeId = ingestEvent(events.next());
+            } while (events.hasNext() && nodeId < lastDescendantNodeId);
```

Comment:
I infer that the assumed stop condition here is that the event stream has produced a node whose ID is exactly equal to `lastDescendantNodeId`? Otherwise `ingestEvent` would consume the first event after `lastDescendantNodeId`, right?

Perhaps an assert to clarify our assumption here?